curl/tests/client_test.cc

```diff
@@ -0,0 +1,405 @@
+        }
+
+        SECTION("cURL should receive the path to the CA certificate specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_ca_cert("cacert");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cacert == "cacert");
+        }
+
+        /*
+         * client_cert and client_key tests
+         */
+        SECTION("Client cert name should be unspecified by default") {
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_cert == "");
+        }
+
+        SECTION("cURL should receive the client cert name specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_client_cert("cert", "key");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_cert == "cert");
+        }
+
+        SECTION("Private keyfile name should be unspecified by default") {
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_key == "");
+        }
+
+        SECTION("cURL should receive the private keyfile name specified in the request") {
+            request test_request {"http://valid.com"};
+            test_client.set_client_cert("cert", "key");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->client_key == "key");
+        }
+    }
+
+    /* Exception handling tests below */
+
+    /*
+    TEST_CASE("curl::client fails to globally initialize libcurl") {
+        SECTION("Client should throw an http_exception") {
+            // curl_global_init()
+            cout << "\nFIRST ERROR TEST\n";
+            curl_fail_init mock_error {global_init_error};
+            REQUIRE_THROWS_AS(mock_client test_client, http_exception);
+        }
+    }
+    */
+
+    TEST_CASE("curl::client exceptions") {
+        SECTION("client fails to initialize a libcurl easy session") {
+            curl_fail_init mock_error {easy_init_error};
+            REQUIRE_THROWS_AS(mock_client test_client, http_exception);
+        }
+
+        SECTION("client fails to perform a cURL request") {
+            mock_client test_client;
```

Comment:
Everything except setting the test_failure_mode and the REQUIRE behavior can be pulled up into the TEST_CASE block if you move the 1st SECTION into its own TEST_CASE. That would make this a lot less verbose.